pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,574 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storageos
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/util/mount"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+	"k8s.io/kubernetes/pkg/volume/util"
+	"k8s.io/kubernetes/pkg/volume/util/volumehelper"
+)
+
+// ProbeVolumePlugins is the primary entrypoint for volume plugins.
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&storageosPlugin{nil}}
+}
+
+type storageosPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &storageosPlugin{}
+var _ volume.PersistentVolumePlugin = &storageosPlugin{}
+var _ volume.DeletableVolumePlugin = &storageosPlugin{}
+var _ volume.ProvisionableVolumePlugin = &storageosPlugin{}
+
+const (
+	storageosPluginName = "kubernetes.io/storageos"
+	storageosDevicePath = "/var/lib/storageos/volumes"
+	defaultAPIAddress   = "tcp://localhost:5705"
+	defaultAPIUser      = "storageos"
+	defaultAPIPassword  = "storageos"
+	defaultAPIVersion   = "1"
+	defaultFSType       = "ext4"
+	defaultNamespace    = "default"
+)
+
+func getPath(uid types.UID, namespace string, volName string, pvName string, host volume.VolumeHost) string {
+	if len(namespace) != 0 && len(volName) != 0 && strings.Count(volName, "~") == 0 {
+		return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(storageosPluginName), namespace+"~"+volName)
+	}
+	return host.GetPodVolumeDir(uid, kstrings.EscapeQualifiedNameForDisk(storageosPluginName), pvName)
+}
+
+func (plugin *storageosPlugin) Init(host volume.VolumeHost) error {
+	plugin.host = host
+	return nil
+}
+
+func (plugin *storageosPlugin) GetPluginName() string {
+	return storageosPluginName
+}
+
+func (plugin *storageosPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	volumeSource, _, err := getVolumeSource(spec)
+	if err != nil {
+		return "", err
+	}
+	return volumeSource.VolumeName, nil
```

Comment:
You should return storageOS namespace/volumename here instead since they are more unique than `volumesource.VolumeName`.